cores/nRF5/main.cpp

```diff
@@ -16,10 +16,30 @@
 #define ARDUINO_MAIN
 #include "Arduino.h"
 
+#if defined(RESET_PIN)
+#include "nrf.h"
+#endif
+
 // Weak empty variant initialization function.
 // May be redefined by variant files.
 void initVariant() __attribute__((weak));
-void initVariant() { }
+void initVariant()
+{
+  #if defined(RESET_PIN)
+  if (((NRF_UICR->PSELRESET[0] & UICR_PSELRESET_CONNECT_Msk) != (UICR_PSELRESET_CONNECT_Connected << UICR_PSELRESET_CONNECT_Pos)) ||
+      ((NRF_UICR->PSELRESET[1] & UICR_PSELRESET_CONNECT_Msk) != (UICR_PSELRESET_CONNECT_Connected << UICR_PSELRESET_CONNECT_Pos))){
+      NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Wen << NVMC_CONFIG_WEN_Pos;
+      while (NRF_NVMC->READY == NVMC_READY_READY_Busy){}
+      NRF_UICR->PSELRESET[0] = RESET_PIN;
+      while (NRF_NVMC->READY == NVMC_READY_READY_Busy){}
+      NRF_UICR->PSELRESET[1] = RESET_PIN;
+      while (NRF_NVMC->READY == NVMC_READY_READY_Busy){}
+      NRF_NVMC->CONFIG = NVMC_CONFIG_WEN_Ren << NVMC_CONFIG_WEN_Pos;
+      while (NRF_NVMC->READY == NVMC_READY_READY_Busy){}
+      NVIC_SystemReset();
+  }
+  #endif
+}
 
```

Comment:
I think this is better suited to go in [wiring.c](https://github.com/sandeepmistry/arduino-nRF5/blob/master/cores/nRF5/wiring.c). In case a variant overrides `initVariant`.